share/completions/gpg2.fish

```diff
@@ -0,0 +1,433 @@
+#
+# Completions for the gpg2 command.
+#
+# This program accepts an rather large number of switches. It allows
+# you to do things like changing what file descriptor errors should be
+# written to, to make gpg2 use a different locale than the one
+# specified in the environment or to specify an alternative home
+# directory.
+
+# Switches related to debugging, switches whose use is not
+# recommended, switches whose behaviour is as of yet undefined,
+# switches for experimental features, switches to make gpg2 compliant
+# to legacy pgp-versions, dos-specific switches, switches meant for
+# the options file and deprecated or obsolete switches have all been
+# removed. The remaining list of completions is still quite
+# impressive.
+
+#
+# Various functions used for dynamic completions
+#
+
+function __fish_complete_gpg2_user_id -d "Complete using gpg2 user ids"
+    # gpg2 doesn't seem to like it when you use the whole key name as a
+    # completion, so we skip the <EMAIL> part and use it a s a
+    # description.
+    # It also replaces colons with \x3a
+    gpg2 --list-keys --with-colon | cut -d : -f 10 | sed -ne 's/\\\x3a/:/g' -e 's/\(.*\) <\(.*\)>/\1'\t'\2/p'
+end
+
+function __fish_complete_gpg2_key_id -d 'Complete using gpg2 key ids'
+    # Use user_id as the description
+    set -l lastid
+    gpg2 --list-keys --with-colons | while read garbage
+        switch $garbage
+            case "uid*"
+                echo $garbage | cut -d ":" -f 10 | sed -e "s/\\\x3a/:/g" | read lastid
+            case "*"
+                echo $garbage | cut -d ":" -f 5 | read fingerprint
+        end
+        printf "%s\t%s\n" $fingerprint $lastid
+    end
+end
```

Comment:
I don't think this function works either - instead of `case "*"`, should it be `case "fpr*"`, and the `printf` moved to the uid case? The local variable will need updating too.

Reply:
Hmm, this seems to work fine for me-- it correctly fetches the keyids in my keyring and offers them as completions. What didn't seem to work for you?

Reply:
The head of `gpg2 --list-keys --with-colons` in GPG 2.2.17 looks like this:

```
tru::1:100:200:3:1:5
pub:u:2048:1:7A67D962D88A709A:1342943306:::u:::scESC::::::23::0:
fpr:::::::::003837986104878835FA516D7A67D962D88A709A:
uid:u::::1459927992::F01DF13E08CDA181385262F6F4D6F4B01A88DE3E::David Adam <zanchey@gmail.com>::::::::::0:
uid:u::::1459927992::08960C4FC514C0116682B1F952653880BF952330::David Adam <zanchey@ucc.gu.uwa.edu.au>::::::::::0:
uid:u::::1459927992::022BDDF556AF93BCB6D8B389A65339CA6B69DB2F::David Adam (keybase.io/zanchey)::::::::::0:
sub:u:2048:1:58FC3A8F0C883551:1342943306::::::e::::::23:
fpr:::::::::2C4C0B1A17510B573F88F9F058FC3A8F0C883551:
sub:u:4096:1:C0B969B2974E888E:1398176859::::::s::::::23:
fpr:::::::::9E1DE0673CCC032919D18526C0B969B2974E888E:
pub:-:4096:1:2A32340CEC8C9492:1398524918:::-:::scESC::::::23::0:
fpr:::::::::03E146CDAF8136680AD566912A32340CEC8C9492:
uid:-::::1398524918::C912CAC6526A647C29FCA77D8D0873EC1A561C93::Keybase.io Merkle Signing (v1) <merkle@keybase.io>::::::::::0:
sub:-:4096:1:8A01CE578080955B:1398524918::::::e::::::23:
fpr:::::::::307BC124E498350024F761DC8A01CE578080955B:
sub:-:4096:1:F43803A349DA99D5:1398524988:1713884988:::::s::::::23:
fpr:::::::::38B848EDBAB432EBDF478992F43803A349DA99D5:
```

and the corresponding output is:

```
0
7A67D962D88A709A

	David Adam <zanchey@gmail.com>
	David Adam <zanchey@ucc.gu.uwa.edu.au>
	David Adam (keybase.io/zanchey)
58FC3A8F0C883551	David Adam (keybase.io/zanchey)
	David Adam (keybase.io/zanchey)
C0B969B2974E888E	David Adam (keybase.io/zanchey)
2A32340CEC8C9492	David Adam (keybase.io/zanchey)
	David Adam (keybase.io/zanchey)
	Keybase.io Merkle Signing (v1) <merkle@keybase.io>
8A01CE578080955B	Keybase.io Merkle Signing (v1) <merkle@keybase.io>
	Keybase.io Merkle Signing (v1) <merkle@keybase.io>
F43803A349DA99D5	Keybase.io Merkle Signing (v1) <merkle@keybase.io>
	Keybase.io Merkle Signing (v1) <merkle@keybase.io>
5AE9196C1DE5F269	Keybase.io Merkle Signing (v1) <merkle@keybase.io>
	Keybase.io Merkle Signing (v1) <merkle@keybase.io>
```

Issues:
* the `tru` line is parsed, but shouldn't be
* my main keyid (7A67D962D) has no description
* my UIDs are printed as descriptions without keyids
* my subkeys are all printed with the last of my UIDs (I don't think there's much point in offering subkey IDs)

This works a bit better:

```
function __fish_complete_gpg2_key_id --description 'Complete using gpg2 key ids'
    set -l keyid
    gpg2 --list-keys --with-colons | while read garbage
        switch $garbage
            case "uid*"
                echo $garbage | cut -d ":" -f 10 | sed -e "s/\\\x3a/:/g" | read uid
                printf "%s\t%s\n" $keyid $uid
            case "pub*"
                echo $garbage | cut -d ":" -f 5 | read keyid
        end
    end
end
```

producing

```
7A67D962D88A709A	David Adam <zanchey@gmail.com>
7A67D962D88A709A	David Adam <zanchey@ucc.gu.uwa.edu.au>
7A67D962D88A709A	David Adam (keybase.io/zanchey)
2A32340CEC8C9492	Keybase.io Merkle Signing (v1) <merkle@keybase.io>
```

Reply:
Ah, I see-- this wasn't obvious to me. Looks like your revised function fixes the issue outlined, though, so I've merged it into my branch. Looks like the duplicate keyid options are automatically eliminated, as well (by `complete`?).

Reply:
Yes, they will be collapsed, which is a bit annoying if they have different descriptions.